clients/apple/Shared/Stateful Logic/BillingService.swift

```diff
@@ -0,0 +1,197 @@
+        do {
+            guard let monthlySubscription = maybeMonthlySubscription else {
+                throw StoreError.noProduct
+            }
+            let result = try await monthlySubscription.purchase(options: purchaseOpt)
+            SKReceiptRefreshRequest().start()
+            
+            switch result {
+            case .success(let verification):
+                while true {
+                    if let receipt = getReceipt() {
+                        makingPurchaseAttempt = true
+                        let transaction = try checkVerified(verification)
+                        
+                        let result = self.core.upgradeAccountAppStore(originalTransactionId: String(transaction.originalID), appAccountToken: accountToken.uuidString.lowercased(), encodedReceipt: receipt)
+                        
+                        switch result {
+                        case .success(_):
+                            await transaction.finish()
+                            makingPurchaseAttempt = false
+                            purchaseResult = .success
+                            return .success
+                        case .failure(let error):
+                            makingPurchaseAttempt = false
+                            DI.errors.handleError(error)
+                            return .failure
+                        }
+                    }
+                }
+            case .pending:
+                purchaseResult = .pending
+                return .pending
+            case .userCancelled:
+                return nil
+            default:
+                return .failure
+            }
+        }
+        catch {
+            print("Couldn't read receipt data with error: " + error.localizedDescription)
+        }
+
+        return .failure
+    }
+    
+    func checkVerified<T>(_ result: VerificationResult<T>) throws -> T {
+        switch result {
+        case .unverified:
+            throw StoreError.failedVerification
+        case .verified(let safe):
+            return safe
+        }
+    }
+    
+    func getReceipt() -> String? {
+        if let appStoreReceiptURL = Bundle.main.appStoreReceiptURL,
+            FileManager.default.fileExists(atPath: appStoreReceiptURL.path) {
+
+            do {
+                return try Data(contentsOf: appStoreReceiptURL, options: .alwaysMapped).base64EncodedString()
+            }
+            catch {
+                return nil
+            }
+        }
+        
+        return nil
+    }
+    
+    func requestProducts() async {
+        do {
+            let storeProducts = try await Product.products(for: [MONTHLY_SUBSCRIPTION_PRODUCT_ID])
+
+            if storeProducts.count == 1 {
+                maybeMonthlySubscription = storeProducts[0]
+            } else {
+                print("No products!")
+            }
+            
+        } catch {
+            print("Failed product request from the App Store server: \(error)")
+        }
+    }
```

Comment:
I wonder if this fails on startup, will something later on act unexpectedly

Reply:
It won't act unexpectedly later. The `maybeMonthlySubscription` is a nullable `Product`. If it is null at the `purchaseProduct` method, then I will simply say "Failed to complete purchase."